core/qt-ble.cpp

```diff
@@ -764,13 +764,50 @@ dc_status_t BLEObject::get_name(char *data, size_t size)
 	return DC_STATUS_SUCCESS;
 }
 
+dc_status_t BLEObject::read_characteristic(const QBluetoothUuid &uuid, char *res, size_t size)
+{
+	auto s = preferredService();
+	if (!s) {
+		return DC_STATUS_NODEVICE;
+	}
+
+	const QList<QLowEnergyCharacteristic> chars = s->characteristics();
+	for (const QLowEnergyCharacteristic& ch : chars) {
+		if (ch.uuid() != uuid) {
+			continue;
+		}
+		report_info("Reading BLE characteristic %s", to_str(uuid).c_str());
+		s->readCharacteristic(ch);
+		QByteArray val = ch.value();
+		if ((size_t)val.size() != size) {
+			report_error("Unexpected read characteristic size (%lld): expected %lld", (long long)val.size(), (long long)size);
+			return DC_STATUS_DATAFORMAT;
+		}
+		report_info("Obtained characteristic value: %s", val.toHex().toStdString().c_str());
+		memcpy(res, val.data(), size);
+		return DC_STATUS_SUCCESS;
+	}
+
+	// Not found
+	report_error("characteristic %s not found", to_str(uuid).c_str());
+	return DC_STATUS_NOACCESS;
+}
+
 dc_status_t qt_ble_ioctl(void *io, unsigned int request, void *data, size_t size)
 {
 	BLEObject *ble = (BLEObject *) io;
+	struct quint128 uuid;
```

Comment:
While variable hoisting exists in older parts of the Subsurface codebase, the current approach is to not do this, and instead declare the variable right before its use, and use the minimum scope for it: https://github.com/subsurface/subsurface/blob/master/CODINGSTYLE.md#coding-conventions

For more complex functions this often greatly improves readability.

Reply:
I agree (and that's even how I coded it first) but I was getting errors due to the switch statement. It turns out you can declare the variables within the switch if you don't initialize them at the same time https://stackoverflow.com/questions/92396/why-cant-variables-be-declared-in-a-switch-statement

I've done that now

Reply:
Yes, that works. Probably the more common way to get around this is to create a block for the case statement that starts with a variable initialisation:

```
case DC_IOCTL_BLE_CHARACTERISTIC_READ: {
[...]
}
case [...]
```

Reply:
Would you prefer the block? I think it’s fine without it.

Reply:
Nah, all good the way it is.